Always enable proper symbolize implementation on Windows

rongjiecomputer opened the pull request:
`Sym*` APIs only rely on PDB files, there is no need to link to debug libraries to use `Sym*` APIs.

`_DEBUG` macro only means the binary is compiled and linked to debug libraries (`/MTd`, `/MDd`, `/LDd`) [0]. Generation of PDB file is controlled by `/Zi` or `/Z7` in compile time and `/debug` in link time. This has no predefined macro for us to detect.

[0]: https://docs.microsoft.com/en-us/cpp/preprocessor/predefined-macros?view=vs-2017

/cc @derekmauro

derekmauro commented:
This sounds like the right thing to do and it sounds like an improvement.

The test also uses the `_DEBUG` macro so that the test doesn't fail if it isn't compiled in debug mode:
https://github.com/abseil/abseil-cpp/blob/0dffca4e36791c7beeda04da10460b534283948a/absl/debugging/symbolize_test.cc#L445

I'll talk to some people and see if we should accept this as is or if we should try to get the test to work without relying on the `_DEBUG` macro.

derekmauro commented:
I spent some time trying to get the tests to pass if I remove the `_DEBUG` condition from https://github.com/abseil/abseil-cpp/blob/0dffca4e36791c7beeda04da10460b534283948a/absl/debugging/symbolize_test.cc#L445 but I've been unsuccessful so far.

We would like to have this tested before this change is committed.

The test passes if I do `bazel test -c dbg absl/debugging:symbolize_test`. If what you are saying is true, I would think `bazel test --copt=/Zi --linkopt=/DEBUG absl/debugging:symbolize_test` would also make the test pass, but this is not the case. I didn't figure out why.

rongjiecomputer commented:
From the test log, it seems like MSVC ICF the `Foo::func` into `nonstatic_func` despite applying with `__declspec(noinline)` (just added in this PR). So I have to use the volatile int hack to prevent ICF.

Tested with `bazel test -c opt --copt=/Z7 --linkopt=/DEBUG //absl/debugging:symbolize_test`.

derekmauro commented on absl/base/attributes.h:
```diff
@@ -125,13 +125,19 @@
 #define ABSL_ATTRIBUTE_ALWAYS_INLINE __attribute__((always_inline))
 #define ABSL_HAVE_ATTRIBUTE_ALWAYS_INLINE 1
+#elif defined(_MSC_VER)
+#define ABSL_ATTRIBUTE_ALWAYS_INLINE __forceinline
```
I'm really sorry about this, but I don't know of any way we are going to be able to accept the changes in `attributes.h` in the short term.

The problem is we have a lot of code that uses `ABSL_ATTRIBUTE_ALWAYS_INLINE`, but MSVC is a lot more picky about the placement of `ABSL_ATTRIBUTE_ALWAYS_INLINE` than other compilers. For example, we have code that places this attribute at the end of the declaration, and this would break that code when it is compiled with MSVC. I'm not talking about just the Abseil project, but literally the billions of lines of Google code that depend on this.

I tried to see how easy it would be to fix all the places in the code that were using this in an incompatible way, but there were too many of them. We should be able to write some kind of clang-based tool to fix them all automatically, but it isn't too high on our priority list.

Anyway, I would like to accept this change some day, but today is not that day :-(

rongjiecomputer replied:
I understand. I will remove this from the PR.

derekmauro reviewed (commented):
We can work around the `attributes.h` problem, but we still need to get the tests to work on our CI.  I asked about `bazel test -c opt --copt=/Z7 --linkopt=/DEBUG //absl/debugging:symbolize_test` just to test the theory that this would work, but ultimately we need it to work when we simply run `bazel test absl/...`, since that is what our CI runs and what our users will run. We will also need to get the equivalent cmake test command to work as well.

I tried adding the compile and link flags to the appropriate BUILD rule, but that doesn't seem sufficient. I guess the whole binary needs the compile flags? I don't know, I'm not an MSVC expert.

Here is what I tried if you want an idea to play with: https://github.com/derekmauro/abseil-cpp/commit/abc0addab545cd19199a41c2aecadadf9bfc0cb1

rongjiecomputer commented:
> I tried adding the compile and link flags to the appropriate BUILD rule, but that doesn't seem sufficient. I guess the whole binary needs the compile flags? I don't know, I'm not an MSVC expert.

Adding `/Z7` only to certain source files is fine, that means only those source files will have symbols included in the PDB as long as final link step has `/DEBUG` flag. I don't see anything wrong with your `BUILD` changes.

I just forked your repo and run the test, it works! Not sure why it is not working for your. I am using Bazel 0.21 + Visual Studio 2017 15.9. I don't think VS version matters. Can you try to use the latest version of Bazel and upload the test failure log?

I will port the changes for CMake too.

derekmauro commented:
> > I tried adding the compile and link flags to the appropriate BUILD rule, but that doesn't seem sufficient. I guess the whole binary needs the compile flags? I don't know, I'm not an MSVC expert.
> 
> Adding `/Z7` only to certain source files is fine, that means only those source files will have symbols included in the PDB as long as final link step has `/DEBUG` flag. I don't see anything wrong with your `BUILD` changes.

That's exactly what I would have thought as well.

> I just forked your repo and run the test, it works! Not sure why it is not working for your. I am using Bazel 0.21 + Visual Studio 2017 15.9. I don't think VS version matters. Can you try to use the latest version of Bazel and upload the test failure log?

```
>bazel version
INFO: Invocation ID: f64bbf7b-15c9-4d31-b12b-c0bb5e155007
Build label: 0.22.0
Build target: bazel-out/x64_windows-opt/bin/src/main/java/com/google/devtools/bu
ild/lib/bazel/BazelServer_deploy.jar
Build time: Mon Jan 28 12:59:27 2019 (1548680367)
Build timestamp: 1548680367
Build timestamp as int: 1548680367
```

```
installationName: VisualStudio/15.8.8+28010.2048
```

```
bazel-bin\absl\debugging\symbolize_test.exe
[==========] Running 3 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 3 tests from Symbolize
[ RUN      ] Symbolize.Basics
absl/debugging/symbolize_test.cc(476): error: Expected equality of these values:

  "win32_nonstatic_func"
  TrySymbolize((void *)(&win32_nonstatic_func))
    Which is: NULL
absl/debugging/symbolize_test.cc(481): error: Value of: static_func_symbol != nu
llptr
  Actual: false
Expected: true
[  FAILED  ] Symbolize.Basics (16 ms)
[ RUN      ] Symbolize.Truncation
absl/debugging/symbolize_test.cc(491): error: Expected equality of these values:

  "win32_nonstatic_func"
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), strlen(kNonStaticFunc)
+ 1)
    Which is: NULL
absl/debugging/symbolize_test.cc(494): error: Expected equality of these values:

  "win32_nonstatic_..."
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), strlen(kNonStaticFunc)
+ 0)
    Which is: NULL
absl/debugging/symbolize_test.cc(497): error: Expected equality of these values:

  "win32_nonstatic..."
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), strlen(kNonStaticFunc)
- 1)
    Which is: NULL
absl/debugging/symbolize_test.cc(499): error: Expected equality of these values:

  "w..."
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), 5)
    Which is: NULL
absl/debugging/symbolize_test.cc(501): error: Expected equality of these values:

  "..."
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), 4)
    Which is: NULL
absl/debugging/symbolize_test.cc(502): error: Expected equality of these values:

  ".."
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), 3)
    Which is: NULL
absl/debugging/symbolize_test.cc(503): error: Expected equality of these values:

  "."
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), 2)
    Which is: NULL
absl/debugging/symbolize_test.cc(504): error: Expected equality of these values:

  ""
  TrySymbolizeWithLimit((void *)(&win32_nonstatic_func), 1)
    Which is: NULL
[  FAILED  ] Symbolize.Truncation (16 ms)
[ RUN      ] Symbolize.SymbolizeWithDemangling
absl/debugging/symbolize_test.cc(510): error: Value of: result != nullptr
  Actual: false
Expected: true
[  FAILED  ] Symbolize.SymbolizeWithDemangling (0 ms)
[----------] 3 tests from Symbolize (32 ms total)

[----------] Global test environment tear-down
[==========] 3 tests from 1 test suite ran. (32 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 3 tests, listed below:
[  FAILED  ] Symbolize.Basics
[  FAILED  ] Symbolize.Truncation
[  FAILED  ] Symbolize.SymbolizeWithDemangling

 3 FAILED TESTS
```

rongjiecomputer committed:
Always enable proper symbolize implementation on Windows

rongjiecomputer commented:
Since `TrySymbolize*` are returning `nullptr`, that means `absl::Symbolize` has failed.

Can you try to modify one line in `symbolize_test.cc`, then run `bazel build -c opt -s //absl/debugging/symbolize_test` and show me the console output? That will show us the exact commands for compiling and linking.

I have ported the BUILD change to CMake too.

derekmauro commented:
It seems to work in `opt` mode for me now (not sure if it didn't before), but it still doesn't work in fastbuild mode (which is the default get when you don't supply `-c <mode>`). I made sure to run `bazel clean --expunge` in between testing in `opt` and `fastbuild` mode.

```
>bazel test -s //absl/debugging:symbolize_test
INFO: Invocation ID: 22d9c8f8-e509-42c3-86b2-846526cd22b1
Loading:
Loading: 0 packages loaded
Analyzing: target //absl/debugging:symbolize_test (0 packages loaded, 0 targets
configured)
INFO: Analysed target //absl/debugging:symbolize_test (1 packages loaded, 16 tar
gets configured).
INFO: Found 1 test target...
[0 / 6] [-----] BazelWorkspaceStatusAction stable-status.txt
SUBCOMMAND: # //absl/debugging:symbolize_test [action 'Compiling absl/debugging/
symbolize_test.cc']
cd C:/users/dmauro/_bazel_dmauro/e3v7ful3/execroot/com_google_absl
  SET INCLUDE=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\T
ools\MSVC\14.15.26726\ATLMFC\include;C:\Program Files (x86)\Microsoft Visual Stu
dio\2017\Community\VC\Tools\MSVC\14.15.26726\include;C:\Program Files (x86)\Wind
ows Kits\NETFXSDK\4.6.1\include\um;C:\Program Files (x86)\Windows Kits\10\includ
e\10.0.17763.0\ucrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\
shared;C:\Program Files (x86)\Windows Kits\10\include\10.0.17763.0\um;C:\Program
 Files (x86)\Windows Kits\10\include\10.0.17763.0\winrt;C:\Program Files (x86)\W
indows Kits\10\include\10.0.17763.0\cppwinrt
    SET PATH=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\To
ols\MSVC\14.15.26726\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Stu
dio\2017\Community\Common7\IDE\VC\VCPackages;C:\Program Files (x86)\Microsoft Vi
sual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\
Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExt
ensions\Microsoft\TeamFoundation\Team Explorer;C:\Program Files (x86)\Microsoft
Visual Studio\2017\Community\MSBuild\15.0\bin\Roslyn;C:\Program Files (x86)\Micr
osoft Visual Studio\2017\Community\Team Tools\Performance Tools\x64;C:\Program F
iles (x86)\Microsoft Visual Studio\2017\Community\Team Tools\Performance Tools;C
:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollectionTool
s\\x64;C:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollec
tionTools\;C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.6.1
Tools\x64\;C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64;C:\Progra
m Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Microsoft Visual St
udio\2017\Community\\MSBuild\15.0\bin;C:\windows\Microsoft.NET\Framework64\v4.0.
30319;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\
;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\Tools\;;C
:\windows\system32;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community
\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin;C:\Program Files (x86)\M
icrosoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\CMa
ke\Ninja
    SET PWD=/proc/self/cwd
    SET TEMP=C:\Users\dmauro\AppData\Local\Temp\3
    SET TMP=C:\Users\dmauro\AppData\Local\Temp\3
  C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14
.15.26726/bin/HostX64/x64/cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WIN
NT=0x0601 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /bigobj /Zm500 /E
Hsc /wd4351 /wd4291 /wd4250 /wd4996 /I. /Ibazel-out/x64_windows-fastbuild/genfil
es /Ibazel-out/x64_windows-fastbuild/bin /Iexternal/com_google_googletest /Ibaze
l-out/x64_windows-fastbuild/genfiles/external/com_google_googletest /Ibazel-out/
x64_windows-fastbuild/bin/external/com_google_googletest /Iexternal/bazel_tools
/Ibazel-out/x64_windows-fastbuild/genfiles/external/bazel_tools /Ibazel-out/x64_
windows-fastbuild/bin/external/bazel_tools /Iexternal/com_google_googletest/goog
lemock /Ibazel-out/x64_windows-fastbuild/genfiles/external/com_google_googletest
/googlemock /Ibazel-out/x64_windows-fastbuild/bin/external/com_google_googletest
/googlemock /Iexternal/com_google_googletest/googlemock/include /Ibazel-out/x64_
windows-fastbuild/genfiles/external/com_google_googletest/googlemock/include /Ib
azel-out/x64_windows-fastbuild/bin/external/com_google_googletest/googlemock/inc
lude /Iexternal/com_google_googletest/googletest /Ibazel-out/x64_windows-fastbui
ld/genfiles/external/com_google_googletest/googletest /Ibazel-out/x64_windows-fa
stbuild/bin/external/com_google_googletest/googletest /Iexternal/com_google_goog
letest/googletest/include /Ibazel-out/x64_windows-fastbuild/genfiles/external/co
m_google_googletest/googletest/include /Ibazel-out/x64_windows-fastbuild/bin/ext
ernal/com_google_googletest/googletest/include /D__CLANG_SUPPORT_DYN_ANNOTATION_
_ /showIncludes /MD /Od /Z7 /wd4117 -D__DATE__="redacted" -D__TIMESTAMP__="redac
ted" -D__TIME__="redacted" /W3 /wd4005 /wd4068 /wd4180 /wd4244 /wd4267 /wd4800 /
DNOMINMAX /DWIN32_LEAN_AND_MEAN /D_CRT_SECURE_NO_WARNINGS /D_SCL_SECURE_NO_WARNI
NGS /D_ENABLE_EXTENDED_ALIGNED_STORAGE /wd4018 /wd4101 /wd4503 /Z7 /Fobazel-out/
x64_windows-fastbuild/bin/absl/debugging/_objs/symbolize_test/symbolize_test.obj
 /c absl/debugging/symbolize_test.cc
[4 / 6] Compiling absl/debugging/symbolize_test.cc; 1s local
SUBCOMMAND: # //absl/debugging:symbolize_test [action 'Linking absl/debugging/sy
mbolize_test.exe']
cd C:/users/dmauro/_bazel_dmauro/e3v7ful3/execroot/com_google_absl
  SET LIB=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools
\MSVC\14.15.26726\ATLMFC\lib\x64;C:\Program Files (x86)\Microsoft Visual Studio\
2017\Community\VC\Tools\MSVC\14.15.26726\lib\x64;C:\Program Files (x86)\Windows
Kits\NETFXSDK\4.6.1\lib\um\x64;C:\Program Files (x86)\Windows Kits\10\lib\10.0.1
7763.0\ucrt\x64;C:\Program Files (x86)\Windows Kits\10\lib\10.0.17763.0\um\x64;
    SET PATH=C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\To
ols\MSVC\14.15.26726\bin\HostX64\x64;C:\Program Files (x86)\Microsoft Visual Stu
dio\2017\Community\Common7\IDE\VC\VCPackages;C:\Program Files (x86)\Microsoft Vi
sual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\
Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\CommonExt
ensions\Microsoft\TeamFoundation\Team Explorer;C:\Program Files (x86)\Microsoft
Visual Studio\2017\Community\MSBuild\15.0\bin\Roslyn;C:\Program Files (x86)\Micr
osoft Visual Studio\2017\Community\Team Tools\Performance Tools\x64;C:\Program F
iles (x86)\Microsoft Visual Studio\2017\Community\Team Tools\Performance Tools;C
:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollectionTool
s\\x64;C:\Program Files (x86)\Microsoft Visual Studio\Shared\Common\VSPerfCollec
tionTools\;C:\Program Files (x86)\Microsoft SDKs\Windows\v10.0A\bin\NETFX 4.6.1
Tools\x64\;C:\Program Files (x86)\Windows Kits\10\bin\10.0.17763.0\x64;C:\Progra
m Files (x86)\Windows Kits\10\bin\x64;C:\Program Files (x86)\Microsoft Visual St
udio\2017\Community\\MSBuild\15.0\bin;C:\windows\Microsoft.NET\Framework64\v4.0.
30319;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\IDE\
;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\Common7\Tools\;;C
:\windows\system32;C:\Program Files (x86)\Microsoft Visual Studio\2017\Community
\Common7\IDE\CommonExtensions\Microsoft\CMake\CMake\bin;C:\Program Files (x86)\M
icrosoft Visual Studio\2017\Community\Common7\IDE\CommonExtensions\Microsoft\CMa
ke\Ninja
    SET PWD=/proc/self/cwd
    SET TEMP=C:\Users\dmauro\AppData\Local\Temp\3
    SET TMP=C:\Users\dmauro\AppData\Local\Temp\3
  C:/Program Files (x86)/Microsoft Visual Studio/2017/Community/VC/Tools/MSVC/14
.15.26726/bin/HostX64/x64/link.exe /nologo /OUT:bazel-out/x64_windows-fastbuild/
bin/absl/debugging/symbolize_test.exe /SUBSYSTEM:CONSOLE /DEBUG /MACHINE:X64 @ba
zel-out/x64_windows-fastbuild/bin/absl/debugging/symbolize_test.exe-2.params /DE
BUG:FASTLINK /INCREMENTAL:NO
SUBCOMMAND: # //absl/debugging:symbolize_test [action 'Testing //absl/debugging:
symbolize_test']
cd C:/users/dmauro/_bazel_dmauro/e3v7ful3/execroot/com_google_absl
  SET JAVA_RUNFILES=bazel-out/x64_windows-fastbuild/bin/absl/debugging/symbolize
_test.exe.runfiles
    SET PATH=C:\msys64\usr\bin;C:\msys64\bin;C:\Perl64\site\bin;C:\Perl64\bin;C:
\Python27\;C:\Python27\Scripts;C:\windows\system32;C:\windows;C:\windows\System3
2\Wbem;C:\windows\System32\WindowsPowerShell\v1.0\;C:\windows\System32\WindowsPo
werShell\v1.0\;C:\ProgramData\GooGet;C:\Program Files (x86)\Google\Cloud SDK\goo
gle-cloud-sdk\bin;C:\windows\System32\WindowsPowerShell\v1.0\;C:\Program Files\G
oogle\Compute Engine\metadata_scripts;C:\Program Files\Google\Compute Engine\sys
prep;C:\ProgramData\chocolatey\bin;C:\Program Files\Git\cmd;C:\windows\system32\
config\systemprofile\.dnx\bin;C:\Program Files\Microsoft DNX\Dnvm\;C:\Program Fi
les\CMake\bin;C:\windows\System32\WindowsPowerShell\v1.0\;C:\Go\bin;C:\Users\dma
uro\go\bin
    SET PYTHON_RUNFILES=bazel-out/x64_windows-fastbuild/bin/absl/debugging/symbo
lize_test.exe.runfiles
    SET RUNFILES_DIR=bazel-out/x64_windows-fastbuild/bin/absl/debugging/symboliz
e_test.exe.runfiles
    SET RUNFILES_MANIFEST_ONLY=1
    SET RUN_UNDER_RUNFILES=1
    SET TEST_BINARY=absl/debugging/symbolize_test.exe
    SET TEST_INFRASTRUCTURE_FAILURE_FILE=bazel-out/x64_windows-fastbuild/testlog
s/absl/debugging/symbolize_test/test.infrastructure_failure
    SET TEST_LOGSPLITTER_OUTPUT_FILE=bazel-out/x64_windows-fastbuild/testlogs/ab
sl/debugging/symbolize_test/test.raw_splitlogs/test.splitlogs
    SET TEST_PREMATURE_EXIT_FILE=bazel-out/x64_windows-fastbuild/testlogs/absl/d
ebugging/symbolize_test/test.exited_prematurely
    SET TEST_SIZE=medium
    SET TEST_SRCDIR=bazel-out/x64_windows-fastbuild/bin/absl/debugging/symbolize
_test.exe.runfiles
    SET TEST_TARGET=//absl/debugging:symbolize_test
    SET TEST_TIMEOUT=300
    SET TEST_TMPDIR=_tmp/2d70fa838f8a62434b2c9930e1f5c140
    SET TEST_UNDECLARED_OUTPUTS_ANNOTATIONS=bazel-out/x64_windows-fastbuild/test
logs/absl/debugging/symbolize_test/test.outputs_manifest/ANNOTATIONS
    SET TEST_UNDECLARED_OUTPUTS_ANNOTATIONS_DIR=bazel-out/x64_windows-fastbuild/
testlogs/absl/debugging/symbolize_test/test.outputs_manifest
    SET TEST_UNDECLARED_OUTPUTS_DIR=bazel-out/x64_windows-fastbuild/testlogs/abs
l/debugging/symbolize_test/test.outputs
    SET TEST_UNDECLARED_OUTPUTS_MANIFEST=bazel-out/x64_windows-fastbuild/testlog
s/absl/debugging/symbolize_test/test.outputs_manifest/MANIFEST
    SET TEST_UNDECLARED_OUTPUTS_ZIP=bazel-out/x64_windows-fastbuild/testlogs/abs
l/debugging/symbolize_test/test.outputs/outputs.zip
    SET TEST_UNUSED_RUNFILES_LOG_FILE=bazel-out/x64_windows-fastbuild/testlogs/a
bsl/debugging/symbolize_test/test.unused_runfiles_log
    SET TEST_WARNINGS_OUTPUT_FILE=bazel-out/x64_windows-fastbuild/testlogs/absl/
debugging/symbolize_test/test.warnings
    SET TEST_WORKSPACE=com_google_absl
    SET TZ=UTC
    SET XML_OUTPUT_FILE=bazel-out/x64_windows-fastbuild/testlogs/absl/debugging/
symbolize_test/test.xml
  C:/msys64/usr/bin/bash.exe -c $0 "$@" external/bazel_tools/tools/test/test-set
up.sh absl/debugging/symbolize_test.exe
[6 / 7] 1 / 1 tests, 1 failed; Testing //absl/debugging:symbolize_test; 0s local

FAIL: //absl/debugging:symbolize_test (see C:/users/dmauro/_bazel_dmauro/e3v7ful
3/execroot/com_google_absl/bazel-out/x64_windows-fastbuild/testlogs/absl/debuggi
ng/symbolize_test/test.log)
Target //absl/debugging:symbolize_test up-to-date:
  C:/users/dmauro/_bazel_dmauro/e3v7ful3/execroot/com_google_absl/bazel-out/x64_
windows-fastbuild/bin/absl/debugging/symbolize_test.exe
INFO: Elapsed time: 3.568s, Critical Path: 2.93s
INFO: 3 processes: 3 local.
INFO: Build completed, 1 test FAILED, 4 total actions
//absl/debugging:symbolize_test                                          FAILED
in 0.4s
  C:/users/dmauro/_bazel_dmauro/e3v7ful3/execroot/com_google_absl/bazel-out/x64_
windows-fastbuild/testlogs/absl/debugging/symbolize_test/test.log

Executed 1 out of 1 test: 1 fails locally.
INFO: Build completed, 1 test FAILED, 4 total actions
```

rongjiecomputer commented:
Fastbuild still works for me without further modification.

I actually did get test failure once when I test the CMake configuration, but wiping CMake build cache and retrying made it work. I don't know what causes these transient test failures to happen.

```
>bazel test //absl/debugging:symbolize_test
INFO: Analysed target //absl/debugging:symbolize_test (20 packages loaded, 555 targets configured).
INFO: Found 1 test target...
Target //absl/debugging:symbolize_test up-to-date:
  C:/users/jie/_bazel_jie/um6asfot/execroot/com_google_absl/bazel-out/x64_windows-fastbuild/bin/absl/debugging/symbolize_test.exe
INFO: Elapsed time: 73.093s, Critical Path: 16.90s
INFO: 40 processes: 40 local.
INFO: Build completed successfully, 53 total actions
//absl/debugging:symbolize_test                                          PASSED in 3.3s

Executed 1 out of 1 test: 1 test passes.
There were tests whose specified size is too big. Use the --test_verbose_timeout
INFO: Build completed successfully, 53 total actions
```

derekmauro commented:
Brief update: I haven't forgotten about this. I'd really like to accept this PR, but we need to figure out why the tests are still failing on our end first. I don't have time to debug this now, and I know you are having trouble reproducing the failure. Hopefully I'll find some time to look into this soon.

zhangxy988 commented:
@derekmauro Have you got some time to look into this?

derekmauro commented:
Sorry, no updates.

rongjiecomputer commented:
Can I just get a CI test? Maybe the issue went away now. It consistently works for me locally with VS2017.

rongjiecomputer commented:
@suertreus Thanks for triggering the CI test. If it is allowed and not too much trouble for you, can you copy the console output of the failing tests here or email it to me so that I can try to fix them?

rongjiecomputer committed:
Remove use of std::aligned_storage

Inspired by How to Hold a T by CJ Johnson

rongjiecomputer commented:
@derekmauro I changed one `std::aligned_storage` to aligned `char` array and added a `GetLastError` printf if `SymFromAddr` fails (this will be removed before merging)

Can you checkout this PR locally and run the following to see if that fixes the issue? If the test still fails, please upload the test log.

```
bazel test //absl/debugging:symbolize_test
```

derekmauro commented on absl/debugging/symbolize_win32.inc:
```diff
@@ -57,12 +57,13 @@ bool Symbolize(const void *pc, char *out, int out_size) {
   symbol->SizeOfStruct = sizeof(SYMBOL_INFO);
   symbol->MaxNameLen = MAX_SYM_NAME;
   if (!SymFromAddr(process, reinterpret_cast<DWORD64>(pc), nullptr, symbol)) {
+    DWORD error = GetLastError();
```
These two lines will need to be removed before submit.

rongjiecomputer replied:
Done

derekmauro commented:
After upgrading some of the Windows test infrastructure this appears to be working. I will see if I can get this merged.

derekmauro commented on absl/debugging/CMakeLists.txt:
```diff
@@ -62,6 +62,9 @@ absl_cc_test(
     "symbolize_test.cc"
   COPTS
     ${ABSL_TEST_COPTS}
+    $<$<BOOL:${MSVC}>:-Z7>
```
From the internal code review: Is there a reason to use `-Z7` and `-DEBUG` here over `/Z7` and `/DEBUG` as you had it in the Bazel file?

rongjiecomputer replied:
For some reason using `/` instead of `-` in CMake generator expression causes weird CMake error, so I just use `-`.

derekmauro commented on absl/debugging/symbolize_test.cc:
```diff
@@ -448,44 +448,74 @@ void ABSL_ATTRIBUTE_NOINLINE TestWithReturnAddress() {
+extern "C" {
+
+__declspec(noinline) void win32_nonstatic_func() {
+  volatile int x = __LINE__;
```
We were wondering if `ABSL_BLOCK_TAIL_CALL_OPTIMIZATION()` is sufficient? What do the extra two lines do?

rongjiecomputer replied:
The two lines are to prevent compiler from folding those functions into one, which I suppose is the original purpose of `ABSL_BLOCK_TAIL_CALL_OPTIMIZATION` as well, but MSVC still do function folding if only using `ABSL_BLOCK_TAIL_CALL_OPTIMIZATION` (tested when comment out those two lines).

derekmauro reviewed (commented):
The internal review had a couple questions for you.

derekmauro committed:
Address Google internal review comments

derekmauro commented:
Merging. Thank you for sticking with this!
